firedrake/assemble_expressions.py

```diff
@@ -1,645 +1,366 @@
-import weakref
-
-import ufl
-from ufl.algorithms import ReuseTransformer
-from ufl.corealg.map_dag import MultiFunction, map_expr_dag
-from ufl.constantvalue import ConstantValue, Zero, IntValue
-from ufl.core.multiindex import MultiIndex
-from ufl.core.operator import Operator
-from ufl.mathfunctions import MathFunction
-from ufl.core.ufl_type import ufl_type as orig_ufl_type
-from ufl import classes
-from collections import defaultdict
 import itertools
+import weakref
+from collections import OrderedDict, defaultdict
 
+import gem
 import loopy
-import pymbolic.primitives as p
+import ufl
+from gem.impero_utils import compile_gem, preprocess_gem
+from gem.node import traversal as gem_traversal
 from pyop2 import op2
-from pyop2.profiling import timed_function
+from pyop2.sequential import Arg
+from tsfc import ufl2gem
+from tsfc.loopy import generate
+from ufl.algorithms.apply_algebra_lowering import LowerCompoundAlgebra
+from ufl.classes import Index, MultiIndex, ConstantValue
+from ufl.corealg.map_dag import map_expr_dags
+from ufl.corealg.multifunction import MultiFunction
+from ufl.corealg.traversal import unique_pre_traversal as ufl_traversal
 
-from firedrake import constant
-from firedrake import function
-from firedrake import utils
+import firedrake
+from firedrake.utils import ScalarType, cached_property, known_pyop2_safe
 
-from functools import singledispatch
 
+def extract_coefficients(expr):
+    return tuple(e for e in ufl_traversal(expr) if isinstance(e, ufl.Coefficient))
```

Comment:
Why do you not use expr.coefficients()?

Reply:
`.coefficients()` isn't implemented for `ufl.Expr` objects. If you mean the `Assign.coefficients` property, that uses this function.

Reply:
This is a micro-optimisation in that I could use `extract_coefficients` but that takes 25µs whereas this takes 17µs on my machine.